drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,230 @@
+	default:
+		return -ENODEV;
+	}
+}
+
+static int ldo_output_enable(const struct device *dev, uint8_t chan, uint8_t enable)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_pconfig *pconfig = config->p->config;
+
+	uint8_t config_bit[3] = {7, 2, 3};
+	uint8_t reg_value = 0;
+
+	i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_DC123_LDO23_CTL, &reg_value);
+
+	if (enable) {
+		reg_value |= BIT(config_bit[chan - 3]);
+	} else {
+		reg_value &= ~BIT(config_bit[chan - 3]);
+	}
+	return i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_DC123_LDO23_CTL, reg_value);
+}
+
+static int ldo_set_voltage(const struct device *dev, uint8_t chan, int32_t init_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_pconfig *pconfig = config->p->config;
+
+	if (init_uv < 1800000 || init_uv > 3300000) {
+		return -EINVAL;
+	}
+
+	uint8_t reg_value = 0;
+	i2c_reg_read_byte_dt(&pconfig->i2c, AXP192_REG_LDO23_OUT_VOL, &reg_value);
+
+	switch (chan) {
+	case AXP192_SOURCE_LDO1:
+		return 0;
+	case AXP192_SOURCE_LDO2:
+		reg_value |= (((init_uv - 180000) / 100000) & 0x0F) << 4;
+		i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_LDO23_OUT_VOL, reg_value);
+		return ldo_output_enable(dev, chan, true);
+	case AXP192_SOURCE_LDO3:
+		reg_value |= (((init_uv - 180000) / 100000) & 0x0F);
+		i2c_reg_write_byte_dt(&pconfig->i2c, AXP192_REG_LDO23_OUT_VOL, reg_value);
+		return ldo_output_enable(dev, chan, true);
+	default:
+		return -ENODEV;
+	}
+}
+
+int regulator_axp192_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	/* MIN and MAX voltage are not used */
+	ARG_UNUSED(min_uv);
+	ARG_UNUSED(max_uv);
+
+	const struct regulator_axp192_config *config = dev->config;
+
+	switch (config->source) {
+	case AXP192_SOURCE_BUCK1:
+		return buck_set_voltage(dev, AXP192_SOURCE_BUCK1, config->init_uv);
+	case AXP192_SOURCE_BUCK2:
+		return buck_set_voltage(dev, AXP192_SOURCE_BUCK2, config->init_uv);
+	case AXP192_SOURCE_BUCK3:
+		return buck_set_voltage(dev, AXP192_SOURCE_BUCK3, config->init_uv);
+	case AXP192_SOURCE_LDO1:
+		return ldo_set_voltage(dev, AXP192_SOURCE_LDO1, config->init_uv);
+	case AXP192_SOURCE_LDO2:
+		return ldo_set_voltage(dev, AXP192_SOURCE_LDO2, config->init_uv);
+	case AXP192_SOURCE_LDO3:
+		return ldo_set_voltage(dev, AXP192_SOURCE_LDO3, config->init_uv);
+	default:
+		return -ENODEV;
+	}
+}
+
+int regulator_axp192_common_init(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return 0;
+}
+
+int regulator_axp192_init(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	int ret = 0;
+
+	if (!device_is_ready(config->p)) {
+		return -ENODEV;
+	}
+
+	ret = regulator_common_init(dev, false);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return ret;
+}
+
+static const struct regulator_driver_api api = {.set_voltage = regulator_axp192_set_voltage};
```

Comment:
I think you should support more of the interface than just set_voltage.
If you use linear_range to define the available voltages, getvoltage, count_voltages and list_voltage will be fairly easy (see the other regulator drivers for inspiration).
Can the regulators be enabled/disabled by SW?

Reply:
> I think you should support more of the interface than just set_voltage.
If you use linear_range to define the available voltages, getvoltage, count_voltages and list_voltage will be fairly easy (see the other regulator drivers for inspiration).

Ok, i will support more interface. 

> Can the regulators be enabled/disabled by SW?

Yes, BUCK1,2,3 and LDO2,3 can be en/dis by softeware.